File: game/game_logic/board.py
# Manages the game board, nodes, and events.
import json
from pathlib import Path
from typing import List, Dict, Any
from .player import Player

class Board:
    """Manages a game board, its nodes, and player movement."""
    def __init__(self, board_id: str):
        self.board_id = board_id
        self.name, self.nodes = self._load_board_data()
        self.player_position = 0

    def _load_board_data(self) -> (str, List[Dict[str, Any]]):
        """Loads the board layout and nodes from the JSON data file."""
        data_path = Path(__file__).parent.parent / "data" / "boards.json"
        with open(data_path, 'r') as f:
            data = json.load(f)
            if self.board_id not in data:
                raise ValueError(f"Board with id '{self.board_id}' not found.")
            board_data = data[self.board_id]
            return board_data['name'], board_data['nodes']

    def move_player(self, player: Player) -> Dict[str, Any] | None:
        """Moves the player to the next node if they have enough energy."""
        if player.energy < 1:
            print("Not enough energy to move!")
            return None

        if self.player_position >= len(self.nodes) - 1:
            print("You have reached the end of the board!")
            return None

        player.energy -= 1
        self.player_position += 1
        current_node = self.nodes[self.player_position]
        print(f"Moved to node {self.player_position + 1}. Event: {current_node['type']}")
        return current_node

    def __str__(self):
        return f"Board: {self.name} ({len(self.nodes)} nodes)"

File: unity_project/Assets/Scripts/Board.cs
using UnityEngine;

/// <summary>
/// Manages a game board, its nodes, and the player's position on it.
/// This is a C# conversion of the Python Board class from the original prototype.
/// It is a plain C# object that holds state and is managed by the GameManager.
/// </summary>
public class Board
{
    public string Name { get; private set; }
    public BoardData Data { get; private set; }

    private int playerPosition = -1; // Start at -1, so the first move is to position 0

    public Board(BoardData boardData)
    {
        Data = boardData;
        Name = boardData.name;
    }

    /// <summary>
    /// Moves the player to the next node on the board if they have enough energy.
    /// </summary>
    /// <param name="player">The player instance, used to check and decrement energy.</param>
    /// <returns>The NodeEvent for the new position, or null if the move is not possible or the board ends.</returns>
    public NodeEvent MovePlayer(Player player)
    {
        if (player.Energy < 1)
        {
            Debug.Log("Not enough energy to move!");
            return null;
        }

        if (playerPosition >= Data.nodes.Count - 1)
        {
            Debug.Log("You have reached the end of the board!");
            return null;
        }

        player.Energy -= 1;
        playerPosition++;

        NodeEvent currentNode = Data.nodes[playerPosition];
        Debug.Log($"Moved to node {playerPosition + 1}. Event: {currentNode.type}");

        return currentNode;
    }
}
